### PR TITLE
[Merged by Bors] - implement reflection for more glam types

### PR DESCRIPTION
**Pull request by makspll**
# Objective

- To implement `Reflect` for more glam types.  

## Solution

insert `impl_reflect_struct` invocations for more glam types. I am not sure about the boolean vectors, since none of them implement `Serde::Serialize/Deserialize`, and the SIMD versions don't have public fields. 
I do still think implementing reflection is useful for BVec's since then they can be incorporated into `Reflect`'ed components and set dynamically even if as a whole + it's more consistent.

## Changelog
Implemented `Reflect` for the following types
 - BVec2
 - BVec3
 - **BVec3A** (on simd supported platforms only)
 - BVec4
 - **BVec4A** (on simd supported platforms only)
 - Mat2
 - Mat3A
 - DMat2
 - Affine2
 - Affine3A
 - DAffine2
 - DAffine3
 - EulerRot

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]

**Comment by makspll** [comment]
Added an implementation for EulerRot too, but that one on top of no `Serialize` and `Deserialize` also doesn't implement `PartialEq`:
```rust
 impl_reflect_value!(EulerRot(Debug, Default));
```

**Comment by makspll** [comment]
Hmm not sure what's happening with the wasm builds

**Comment by makspll** [comment]
I've looked into this and the relevant glam LOC are here: https://github.com/bitshifter/glam-rs/blob/58285f7b127cb93b314e22344bac678b9ba4298f/src/bool.rs#L25-L35.

There is a glam issue about this behaviour here: https://github.com/bitshifter/glam-rs/issues/306,
It do be annoying.

**Comment by mockersf** [comment]
please mention the glam issue in the comment 👍

File: crates/bevy_reflect/src/impls/glam.rs [hunk 1]
```diff
@@ -158,3 +235,12 @@ impl_reflect_value!(DQuat(Debug, PartialEq, Serialize, Deserialize, Default));
 
 impl_from_reflect_value!(Quat);
 impl_from_reflect_value!(DQuat);
+
+impl_reflect_value!(EulerRot(Debug, Default));
+
+// glam type aliases these to the non simd versions when there is no support (this breaks wasm builds for example)
+// ideally it shouldn't do that and there's an issue on glam for this
```
Comment (mockersf) [inline crates/bevy_reflect/src/impls/glam.rs:242]:
```suggestion
// ideally it shouldn't do that and there's an issue on glam for this
// https://github.com/bitshifter/glam-rs/issues/306
```

**Review by mockersf — APPROVED** [review]

**Comment by bitshifter** [comment]
What is the use case for serializing or reflecting the `BVec*` types? The intention of these types in glam is as a temporary value typically created as outputs from the `cmp*` methods for input into `Vec*::select`. These are optimised for SIMD where everything can live in vector registers and branches can be avoided. For scalar types there is a `bool` fall-back. I didn't anticipate people wanting to serialize them, so it would be good to understand why this is desired.

I'm also curious about the use case for reflecting `EulerRot`. `glam` is coordinate system agnostic, so it aims to support any possible coordinate system. AFAIK Bevy has a defined 2D and 3D coordinate systems and I would have thought it would consistently use a single EulerRot order everywhere?

**Comment by makspll** [comment]
As for reflection:

So I think the main argument for reflecting those is consistency, as an example, I am writing `bevy_mod_scripting` and in order to support the bevy API I am wrapping all of the bevy types in [Lua supporting wrappers](https://github.com/makspll/bevy_mod_scripting/blob/4d30ca36f519613e9036bebf8cc89bdaf650300d/bevy_mod_scripting/src/api/wrappers.rs#L11-L16).
Now with the way bevy reflect is heading, I am assuming that anything which is to be available to scripts will implement `Reflect`. Things which don't must be treated differently since they cannot be part of components of resources which are in fact `Reflect`'able, If all of glam implements Reflect, I don't need to worry about this. But even if these types implement Reflect this doesn't mean they won't just be temporaries, it's more of a convenience for treating types uniformly. 

But that's just why I personally prefer this approach. Other than that, I don't have anything specific in mind but I also don't see why these shouldn't be supported as subfields of components or resources. There *may* be a neat use case for boolean vectors or serializing them, why should we limit the non-rust user?  

Same reasoning applies for EulerRot, while bevy has an established preference for coordinates, someone might prefer not to use this and roll their own transforms.

So my argument here is basically, why not?

Now Serialization is not really that big of a deal here, the API difference between platforms is however. I guess if performance would be much worse with the alternative consistent API though, then reflecting these takes less priority than performance of the other glam operations.

**Comment by bitshifter** [comment]
Thanks for the explanation, exposing to script does make a lot of sense. To be clear I'm not opposing this but I did want to understand the why a bit better. The performance question issue is a bit theoretical at the moment, it would require a bit of work to determine if it was a real problem. For example, `BVec*` could go back to containing `u32` components of `0` or `0xffffffff` like they used to in the early glam days, so the type is the same between SIMD and scalar code. But it might be a performance hit when you aren't using SIMD, which for bevy which mostly uses `Vec3` could be an issue. The simpler thing to do would be to have a scalar implementations of `BVec3A` and `BVec4A` instead of aliasing them when SIMD isn't available. Then any perf hit will only be on platforms that don't have a SIMD implementation yet, like aarch64.

I probably should have started this discussion on the glam issue but I was interested in the bevy context :)

**Comment by makspll** [comment]
Yeah no worries, I get why you'd want context! 

The scalar implementation probably sounds like the best solution in the context of bevy, but I have no expertise in this area

**Pull request closed without merging**